### ISSUE TITLE
Only add null guards when explicitly annotated via [NotNull]

### ISSUE THREAD

**Issue by tom-englert** [Fody/NullGuard#67]
We have lot's of code that is already annotated by R# with [NotNull], and anything without an attribute is assumed to "AllowNull".

NullGuard could provide an "explicit" mode, where only items with the [NotNull] attribute get a null guard injected. 

This would be a cool feature for anyone using R# in the "Pessimistic" mode.

**Comment by ulrichb** [comment]
Hi,

There is already a Fody weaver which does this: [controlflow/resharper-checker](https://github.com/controlflow/resharper-checker).

The philosophy behind NullGuard is exactly the opposite: _NotNull by default_ and _explicit CanBeNull/AllowNull_ for specific elements. BTW: This behavior can be brought to R# by using the [ImplicitNullability](https://github.com/ulrichb/ImplicitNullability) extension.

**Comment by tom-englert** [comment]
It could have saved me some work if this would be listed somewhere ;-)

**Comment by cezarypiatek** [comment]
Is it possible to add switch something like that below?

```xml
<?xml version="1.0" encoding="utf-8" ?>
<Weavers>
  <NullGuard ExplcitAnnotation="true" />
</Weavers>
```

**Comment by tom-englert** [comment]
@cezarypiatek see #68, a beta version is at https://github.com/tom-englert/NullGuard/tree/ExplicitMode2

**Comment by SimonCropp** [comment]
I will look at merging and releasing this on the weekend

**Comment by tom-englert** [comment]
@SimonCropp #68 is outdated, I only kept it open because of the discussions going on there, so people can see there is some progress ongoing.
The new branch is not ready for a PR yet.
Also I'm still not sure whether to merge this, or to create a new spin off, e.g. "ExplicitNullGuard", that works out of the box after installation without having to configure every weaver in every project.
What's your opinion about this? We could even go for both options, what would make synchronizing the share code base easier.

**Comment by SimonCropp** [comment]
@tom-englert perhaps easier if i add u as a contributor?

**Comment by tom-englert** [comment]
Sounds good :smile: 
I'm going to finish the explicit mode in my branch, then come back to you to discuss further strategies.

**Comment by cezarypiatek** [comment]
@tom-englert support for explicit [NotNull] will be invaluable for people using Resharper. But there is one more feature that is not supported by NullGuard and it is specific only for Resharper - attribute inheritance from interface.  For example

```cs
public interface IMyType{
  void Execute([NotNull]OtherType x );
}

public class MyType: IMyType
{
  void Execute(OtherType x ){
    throw new NotImplementedException();
  }
}
```

C# compiler does not add attributes from parameters of method declaration to parameters of method definition. However Resharpen static analyzer is using attributes from method declaration to validate method definition. This prevents contract duplication and inconsistency between definition and declaration. It would be very helpful if this rule could be respected by NullGuard in reference to Resharper [NotNull] attribute. If this approach is acceptable I could help with implementation (I have already started some research on local copy).

What do you think about that?

**Comment by tom-englert** [comment]
@cezarypiatek I'm just working on this 😄 (see my last commits, it's already checking for base classes, interfaces are under construction)
It also needs to read external annotations...

**Comment by tom-englert** [comment]
Fixed by #88
